src/Microsoft.AspNet.Server.Kestrel/Server.cs

```diff
@@ -0,0 +1,120 @@
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+            if (httpContextFactory == null)
+            {
+                throw new ArgumentNullException(nameof(httpContextFactory));
+            }
+
+            _applicationLifetime = applicationLifetime;
+            _logger = logger;
+            Features = features;
+            _httpContextFactory = httpContextFactory;
+        }
+
+        public void Start(RequestDelegate requestDelegate)
+        {
+            _disposables = new Stack<IDisposable>();
+
+            try
+            {
+                var information = (KestrelServerInformation)Features.Get<IKestrelServerInformation>();
+                var dateHeaderValueManager = new DateHeaderValueManager();
+                var engine = new KestrelEngine(new ServiceContext
+                {
+                    AppLifetime = _applicationLifetime,
+                    Log = new KestrelTrace(_logger),
+                    HttpContextFactory = _httpContextFactory,
+                    DateHeaderValueManager = dateHeaderValueManager,
+                    ConnectionFilter = information.ConnectionFilter,
+                    NoDelay = information.NoDelay
+                });
+
+                _disposables.Push(engine);
+                _disposables.Push(dateHeaderValueManager);
+
+                if (information.ThreadCount < 0)
+                {
+                    throw new ArgumentOutOfRangeException(nameof(information.ThreadCount),
+                        information.ThreadCount,
+                        "ThreadCount cannot be negative");
+                }
+
+                engine.Start(information.ThreadCount == 0 ? 1 : information.ThreadCount);
+                bool atLeastOneListener = false;
+
+                foreach (var address in information.Addresses)
+                {
+                    var parsedAddress = ServerAddress.FromUrl(address);
+                    if (parsedAddress == null)
+                    {
+                        throw new FormatException("Unrecognized listening address: " + address);
+                    }
+                    else
+                    {
+                        atLeastOneListener = true;
+                        _disposables.Push(engine.CreateServer(
+                            parsedAddress,
+                            requestDelegate));
+                    }
+                }
+
+                if (!atLeastOneListener)
+                {
+                    throw new InvalidOperationException("No recognized listening addresses were configured.");
+                }
+            }
+            catch
+            {
+                Dispose();
+                throw;
+            }
+        }
+
+        public void Dispose()
+        {
+            if (_disposables != null)
+            {
+                foreach (var disposable in _disposables)
+                {
+                    disposable.Dispose();
+                }
```

Comment:
Clean the _disposables afterwards to prevent objects been disposed twice. It happens when `Start` is finished and `Dispose` is invoked again.